Mixpanel/AutomaticEvents.m

```diff
@@ -0,0 +1,158 @@
+}
+
+static NSTimeInterval _appStartTime;
++ (NSTimeInterval)appStartTime { return _appStartTime; }
++ (void)setAppStartTime:(NSTimeInterval)appStartTime { _appStartTime = appStartTime; }
+
+__attribute__((constructor))
+static void initialize_appStartTime() {
+    if (AutomaticEvents.appStartTime == 0) {
+        AutomaticEvents.appStartTime = [[NSDate date] timeIntervalSince1970];
+    }
+}
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self) {
+        awaitingTransactions = [[NSMutableDictionary alloc] init];
+        defaults = [[NSUserDefaults alloc] initWithSuiteName:@"Mixpanel"];
+        sessionLength = 0;
+        sessionStartTime = 0;
+        self.minimumSessionDuration = 10000;
+        self.maximumSessionDuration = UINT64_MAX;
+    }
+    return self;
+}
+
+- (void)initializeEvents:(MixpanelPeople *)peopleInstance {
+    people = peopleInstance;
+    NSString *firstOpenKey = @"MPFirstOpen";
+    if (defaults != nil && ![defaults boolForKey:firstOpenKey]) {
+        if (![self isExistingUser]) {
+            [self.delegate track:@"$ae_first_open" properties:nil];
+            [people setOnce:@{@"$ae_first_app_open_date": [NSDate date]}];
+        }
+        [defaults setBool:TRUE forKey:firstOpenKey];
+        [defaults synchronize];
+    }
+
+    NSDictionary* infoDict = [NSBundle mainBundle].infoDictionary;
+    if (defaults != nil && infoDict != nil) {
+        NSString* appVersionKey = @"MPAppVersion";
+        NSString* appVersionValue = infoDict[@"CFBundleShortVersionString"];
+        NSString* savedVersionValue = [defaults stringForKey:appVersionKey];
+        if (appVersionValue != nil && savedVersionValue != nil && ![appVersionValue isEqualToString:savedVersionValue]) {
+            [self.delegate track:@"$ae_updated" properties:@{@"$ae_updated_version": appVersionValue}];
+            [defaults setObject:appVersionValue forKey:appVersionKey];
+            [defaults synchronize];
+        } else if (savedVersionValue == nil) {
+            [defaults setObject:appVersionValue forKey:appVersionKey];
+            [defaults synchronize];
+        }
+    }
+
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(appWillResignActive:)
+                                                 name:UIApplicationWillResignActiveNotification
+                                               object:nil];
+
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(appDidBecomeActive:)
+                                                 name:UIApplicationDidBecomeActiveNotification
+                                               object:nil];
+
+    [[SKPaymentQueue defaultQueue] addTransactionObserver:self];
+
+}
+
+- (void)appWillResignActive:(NSNotification *)notification {
+    sessionLength = [self roundOneDigit:[[NSDate date] timeIntervalSince1970] - sessionStartTime];
```

Comment:
i think `sessionStartTime` is zero the first time this line is hit and only set properly when the app became active.